WalletWasabi/Blockchain/Blocks/SmartHeaderChain.cs

```diff
@@ -0,0 +1,171 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using WalletWasabi.Bases;
+using WalletWasabi.Helpers;
+using WalletWasabi.Models;
+
+namespace WalletWasabi.Blockchain.Blocks
+{
+	/// <summary>
+	/// High performance chain index and cache.
+	/// </summary>
+	public class SmartHeaderChain : NotifyPropertyChangedBase
+	{
+		private uint _tipHeight;
+		private SmartHeader _tip;
+		private uint256 _tipHash;
+		private uint _serverTipHeight;
+		private int _hashesLeft;
+		private int _hashesCount;
+
+		private SortedDictionary<uint, SmartHeader> Chain { get; }
+		private object Lock { get; }
+
+		public SmartHeader Tip
+		{
+			get => _tip;
+			private set => RaiseAndSetIfChanged(ref _tip, value);
+		}
+
+		public uint TipHeight
+		{
+			get => _tipHeight;
+			private set => RaiseAndSetIfChanged(ref _tipHeight, value);
+		}
+
+		public uint256 TipHash
+		{
+			get => _tipHash;
+			private set => RaiseAndSetIfChanged(ref _tipHash, value);
+		}
+
+		public uint ServerTipHeight
+		{
+			get => _serverTipHeight;
+			private set => RaiseAndSetIfChanged(ref _serverTipHeight, value);
+		}
+
+		public int HashesLeft
+		{
+			get => _hashesLeft;
+			private set => RaiseAndSetIfChanged(ref _hashesLeft, value);
+		}
+
+		public int HashCount
+		{
+			get => _hashesCount;
+			private set => RaiseAndSetIfChanged(ref _hashesCount, value);
+		}
+
+		public SmartHeaderChain()
+		{
+			Chain = new SortedDictionary<uint, SmartHeader>();
+			Lock = new object();
+		}
+
+		public void AddOrReplace(SmartHeader header)
+		{
+			lock (Lock)
+			{
+				var lastHeader = Chain.LastOrDefault().Value;
+				if (lastHeader is { })
+				{
+					if (lastHeader.BlockHash != header.PrevHash)
+					{
+						throw new InvalidOperationException($"Header doesn't point to previous header. Actual: {lastHeader.PrevHash}. Expected: {header.PrevHash}.");
+					}
+
+					if (lastHeader.Height != header.Height - 1)
+					{
+						throw new InvalidOperationException($"Header height isn't one more than the previous header height. Actual: {lastHeader.Height}. Expected: {header.Height - 1}.");
+					}
+				}
+
+				Chain.AddOrReplace(header.Height, header);
+				SetTip(header);
+			}
+		}
+
+		private void SetTip(SmartHeader header)
+		{
+			Tip = header;
+			TipHeight = header.Height;
+			TipHash = header.BlockHash;
+			HashCount = Chain.Count;
+			SetHashesLeft();
+		}
+
+		private void SetHashesLeft()
+		{
+			HashesLeft = (int)Math.Max(0, (long)ServerTipHeight - TipHeight);
+		}
+
+		public void RemoveLast()
+		{
+			lock (Lock)
+			{
+				if (Chain.Any())
+				{
+					Chain.Remove(Chain.Keys.Max());
```

Comment:
Shouldn't this be `Chain.Remove(Chain.Last.Key)`?

Reply:
Done.